Introduce Xsight X2-Based Device for SONiC using Accton platform

deran1980 opened the pull request:
<!--
     Please make sure you've read and understood our contributing guidelines:
     https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

     ** Make sure all your commits include a signature generated with `git commit -s` **

     If this is a bug fix, make sure your description includes "fixes #xxxx", or
     "closes #xxxx" or "resolves #xxxx"

     Please provide the following information:
-->

#### Why I did it

Adding a new device platform support in sonic-buildimage for Xsight Labs devices (based on Accton platforms).
The new device platform name is ES9618XX and currently supports a single SKU O16x400G and it is based on the latest
chip from Xsight - X2.

##### Work item tracking
- Microsoft ADO **(number only)**:
<!-- Replace with your actual ADO work item number -->


#### How I did it

- Added all Accton-relevant platform drivers and code.
- Uploaded Xsight-based platform support 
- Uploaded relevant precompiled Debian packages to the Xsight SONiC repository.

#### How to verify it

Run the compiled SONiC image for the Xsight platform on a Xsight Labs switch and verify successful boot and functionality using Accton ES9618 plaftorm based on the X2 chip.

<!--
If PR needs to be backported, then the PR must be tested against the base branch and the earliest backport release branch and provide tested image version on these two branches. For example, if the PR is requested for master, 202211 and 202012, then the requester needs to provide test results on master and 202012.
-->

#### Which release branch to backport (provide reason below if selected)

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

<!--
- Please provide tested image version
- e.g.
- [x] 20201231.100
-->

#### Description for the changelog

In order to build a SONiC image for the Xsight platform, use the following commands:

```bash
SONIC_BUILD_JOBS=8 SONIC_DPKG_CACHE_METHOD=none make target/sonic-xsight.bin

aparkhomenko-xsight committed:
[Xsight] Migrate docker containers to Debian Bookworm

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] Add support for Accton ES9618XX platform

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>
Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

MichaelZaidman committed:
[ES9618XX] Fix SONiC CI Semgrep scan dangerous-system-call-audit issue

Fixed SONiC CI Semgrep static analyzer scan dangerous-system-call-audit issue:

python.lang.security.audit.dangerous-system-call-audit.dangerous-system-call-audit
Found dynamic content used in a system call. This is dangerous if external data can reach this
function call because it allows a malicious actor to execute commands. Use the 'subprocess' module
instead, which is easier to use without accidentally exposing a command injection vulnerability.
Details: https://sg.run/2WL0

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

MichaelZaidman committed:
[ES9618XX] Fix SONiC CI Semgrep scan subprocess-shell-true issue

This commit fixes the security warning from the SONiC CI Semgrep static analyzer:

python.lang.security.audit.subprocess-shell-true.subprocess-shell-true
Found 'subprocess' function 'Popen' with 'shell=True'. This is dangerous because this call will
spawn the command using a shell process. Doing so propagates current shell settings and variables,
which makes it much easier for a malicious actor to execute commands. Use 'shell=False' instead.
Details: https://sg.run/J92w

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

MichaelZaidman committed:
[ES9618XX] Update path to artifact repository

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

mssonicbld commented:
/azp run Azure.sonic-buildimage

aparkhomenko-xsight committed:
[ES9618XX] remove copp_custom.json

copp_custom.json was used in the past and not needed any more

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

MichaelZaidman commented on device/accton/x86_64-accton_es9618xx-r0/custom.json:
Thanks for the suggestion! That makes sense - with the enhancement in [#20826](https://github.com/sonic-net/sonic-buildimage/pull/20826), adding these settings to `device/accton/x86_64-accton_es9618xx-r0/platform.json` seems appropriate. I'll update the PR accordingly.

mssonicbld commented:
/azp run Azure.sonic-buildimage

aparkhomenko-xsight commented on device/accton/x86_64-accton_es9618xx-r0/Accton-ES9618XX-O16x400G/copp_custom.json:
This file is not needed any more, so I deleted it

CharlieChenEC commented on device/accton/x86_64-accton_es9618xx-r0/fast-reboot_plugin:
```diff
@@ -0,0 +1,24 @@
+	fi
+fi
+
+/opt/xplt/utils/es9618x_reset_x2.sh
```
According to the design of fast-reboot (https://github.com/sonic-net/SONiC/blob/master/doc/fast-reboot/fastreboot.pdf), the data plane remains operational when this file is invoked. I'm wondering if the execution of "es9618x_reset_x2.sh" will stop the data plane from working?

MichaelZaidman replied:
Yes, the script does stop the data plane earlier than expected.

This is a temporary workaround implemented by our engineer to pass the fast-reboot test in the SONiC management test suite:
[sonic-mgmt test_reboot.py#L213](https://github.com/sonic-net/sonic-mgmt/blob/master/tests/platform_tests/test_reboot.py#L213),
which was consistently failing with the following error:
AssertionError: Not all transceivers are detected or interfaces are up in 300 seconds

Here’s the related internal commit message for context:

> SON-1984: es9618xx: fast-reboot_plugin: fix interfaces oper state after running fast-reboot
> Added a fast-reboot_plugin script to unload the XPCI module and reset the chip, as specified in
> Clause 6 of the [Fast-Reboot implementation details](https://github.com/sonic-net/SONiC/wiki/Fast-Reboot#implementation-details).
> Without this script, interfaces remained in the 'down' state after a fast reboot, as shown by
> the show interfaces description command.
> 
> The addition of this script allowed the fast-reboot test to pass by ensuring the interfaces
> transitioned to the 'up' state, thereby avoiding a timeout in the test.

To clarify: full fast-reboot support is not yet implemented. We plan to address it properly in a future PR with a design that aligns with the official fast-reboot flow and minimizes data plane disruption.

CharlieChenEC replied:
Understood.Thanks for your explanation.

MichaelZaidman committed:
[ES9618XX] Use platform.json for ASIC sensor polling configuration

With the enhancement introduced in PR #20826 -
"[asic_sensors] Generate the ASIC sensor polling configuration
based on the platform.json" - we can now define ASIC sensor polling
settings directly in platform.json.

This change removes the need for a custom.json override and simplifies
platform-specific configuration by centralizing the polling interval
and status settings under platform.json.

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

MichaelZaidman commented on platform/xsight/sonic-platform-modules-accton/debian/changelog:
```diff
@@ -0,0 +1,6 @@
+sonic-accton-platform-modules (1.1) unstable; urgency=low
+
+  * Add support for Accton ES9632XX
```
Good catch, thanks! You're absolutely right — it should be ES9618XX. I'll correct it in the next update.

MichaelZaidman replied:
Fixed in https://github.com/sonic-net/sonic-buildimage/pull/22622/commits/269112cbe100ced577fe0c46d73f963b1135362f commit.

MichaelZaidman commented on platform/xsight/sonic-platform-modules-accton/common/udev/70-persistent-net.rules:
Thanks for spotting this - it’s actually a leftover from a previous project and isn’t needed anymore. I’ll remove it in the next update.

MichaelZaidman replied:
Removed in https://github.com/sonic-net/sonic-buildimage/pull/22622/commits/9b47b5232c65bcaecea1b3f7256ad1853907f686 commit.

MichaelZaidman commented on platform/xsight/files/link/70-xsight-eth10g0.link:
You're right - these files are platform-specific, and moving their installation to the platform's Debian post-install script makes sense.

The reason we introduced the 70-xsight-eth10g0.link file is to address an intermittent issue where Intel's 10G hardware is enumerated on the PCI bus before the 1G hardware. In such cases, the ice driver is probed before the igb driver, and the system incorrectly assigns eth0 to the 10G interface. This happens in roughly 10% of reboots.

To avoid this misassignment, we use systemd.link file, which is processed by systemd-udevd earlier than traditional udev rules. This ensures a consistent and deterministic interface naming, regardless of enumeration order.

Thanks for pointing it out - we'll move this logic into the platform’s Debian post-install script accordingly.

MichaelZaidman replied:
Done in https://github.com/sonic-net/sonic-buildimage/pull/22622/commits/269112cbe100ced577fe0c46d73f963b1135362f commit.

MichaelZaidman committed:
[ES9618XX] Install systemd .link files via platform Debian package

We use .link files to address an intermittent issue where Intel's
10G hardware is enumerated on the PCI bus before the 1G hardware.
In such cases, the ice driver is probed before the igb driver,
and the system incorrectly assigns eth0 to the 10G interface.
This happens in roughly 10% of reboots.

To avoid this misassignment, we use systemd.link files, that are
processed by systemd-udevd earlier than traditional udev rules.
This ensures a consistent and deterministic interface naming,
regardless of enumeration order.

These .link files match interfaces by their PCI device addresses,
which may vary across different hardware boards.

Previously, the files were located in platform/xsight/files/link/
and copied to $FILESYSTEM_ROOT/lib/systemd/network/ for all Xsight
platforms via the sonic_debian_extension.j2 template.

This commit relocates the .link files to their respective platform-
specific directories, so they can be installed directly by the
corresponding platform Debian package.

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

MichaelZaidman committed:
[ES9618XX] Clean up obsolete persistent net udev rules

Remove the unused udev rule file:
platform/xsight/sonic-platform-modules-accton/common/udev/70-persistent-net.rules

This file is a remnant from an earlier project and is no longer relevant
to the ES9618XX platform. Its presence serves no functional purpose and may
lead to confusion or unintended behavior if mistakenly modified or applied.

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

mssonicbld commented:
/azp run Azure.sonic-buildimage

CharlieChenEC reviewed (approved):
All of my review comment has been revised. I think the PR is ready for merge.

prgeor commented on platform/xsight/xpci/Makefile:
```diff
@@ -0,0 +1,34 @@
+.SHELLFLAGS += -e
+
+XPCI_DEB = xpci-modules-$(KVERSION)_$(XPCI_VERSION)_amd64.deb
+XPCI_URL_PREFIX = "https://raw.githubusercontent.com/xsightlabs/SONiC/main/amd64/kernel"
```
@deran1980 Why this is not part of sonic-linux-kernel? @saiarcot895 for vs.

saiarcot895 replied:
If I'm reading this right, this suggests that the kernel driver is precompiled, and that precompiled version is being downloaded here. That will not work for SONiC's purposes. If there is a kernel version change, then the kernel modules will need to be recompiled separately from all of the other kernel modules.

ktsym replied:
Hi @saiarcot895. Ok, we will prepare a DKMS-based Debian package that includes the driver sources and compiles automatically for each kernel version. To our understanding, this approach aligns with SONiC requirements and ensures independence from any specific kernel build.

MichaelZaidman committed:
Enable xcvrd_cmis_mgr transceiver management

Originally, we used the /opt/xplt/utils/xcvrs tools in our lab for
SI characterization of modules on our Ubuntu-based XsightOS NOS.
When we began working with SONiC, these tools helped fill the gap
caused by limited CMIS and SI configuration support in earlier versions.

Given their reliability across different environments, we extended
their use to the SONiC 202311 release and, eventually, into this PR.
This has been a temporary measure, and we’re now aligning with the
community's standard approach moving forward.

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

saiarcot895 commented on platform/xsight/docker-syncd-xsight/Dockerfile.j2:
```diff
@@ -0,0 +1,31 @@
+{% from "dockers/dockerfile-macros.j2" import install_debian_packages, install_python_wheels, copy_files %}
+FROM docker-config-engine-bullseye-{{DOCKER_USERNAME}}:{{DOCKER_USERTAG}}
```
Why is bullseye being used here? Why is bookworm not being used?

aparkhomenko-xsight replied:
Done in https://github.com/sonic-net/sonic-buildimage/pull/22622/commits/d01d74512881bf67c34f73d216f00c64e21db705 commit.

aparkhomenko-xsight committed:
[Xsight] remove makefiles for KVM image

- Why I did it
This files is used internally and not needed for publishing

- How I did it
Modified corresponding Makefiles and remove unneeded files

- How to verify it
Make sure compilation and basic tests are successful

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] update packages URLs and version

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] Update XPCI driver to use DKMS infrastructure

- Why I did it
To always build drivers from sources and don't use pre-built drivers

- How I did it
DKMS modules xpci-dkms_0.7.0_amd64.deb is saved and then used to build
the binary module

- How to verify it
Make sure compilation and basic tests are successful

BLDENV=bookworm make -f Makefile.work \
	target/debs/bookworm/xpci-modules-6.1.0-29-2-amd64_0.7.0_amd64.deb

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] Update .dep files for xsai, xpci and xplt

- Why I did it
Changes were advised during PR review

- How I did it
Hashes for builds with DPKG cache are not based on binaries which are
downloaded
Only related source code files are used for generation of hashes

- How to verify it
Make sure compilation and basic tests are successful

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

mssonicbld commented:
/azp run Azure.sonic-buildimage

MichaelZaidman commented on device/accton/x86_64-accton_es9618xx-r0/pmon_daemon_control.json:
```diff
@@ -0,0 +1,4 @@
+{
+    "skip_xcvrd_cmis_mgr": true
```
Done in https://github.com/sonic-net/sonic-buildimage/pull/22622/commits/5f3a3e0a3c23898edde3de5d533534f0cc443938 commit.

aparkhomenko-xsight committed:
[Xsight] fix .dep files for xsai amd xpci

- Why I did it
XPCI_FLAGS and SAI_FLAGS are not needed any more after removing
of wget calls
DEP_FILES must be fixed for xsight-sai.dep

- How I did it
XPCI_FLAGS and SAI_FLAGS have been removed from .dep files
DEP_FILES has been fixed for xsight-sai.dep

- How to verify it
Make sure compilation and basic tests are successful

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

MichaelZaidman commented on platform/xsight/xplt.mk:
```diff
@@ -0,0 +1,9 @@
+# Add platform specific tools
+
+ACCTON_ES9618XX_XPLT_PLATFORM_MODULE = xplt-es9618xx-sonicnos_3.3.3-51-g1a65a01_amd64.deb
+ES9618XX_XPLT_SRV = "https://raw.githubusercontent.com/xsightlabs/SONiC/main/amd64/utils/"
```
Hi @saiarcot895,

Thank you again for your valuable feedback.

Following your comments, we held an internal discussion to evaluate the available approaches for managing binary artifacts. We specifically considered the two options you proposed, along with current practices in similar upstream contributions.

We reviewed two common methods:

1. **Binaries as Regular Git Files** (as used by Marvell)
   *Pros:*

   * Simplicity in implementation and usage
   * Fine-grained control: each binary can evolve independently
   * Straightforward per-file tracking and updating
  
   *Cons:*
   * Rapid repository size growth
   * Lacks native release/version management

2. **GitHub Releases for Binary Artifacts** (as used by Mellanox)
   *Pros:*

   * Purpose-built for managing binary deliverables
   * Clear versioning and release visibility
   * Suitable for formal artifact grouping
 
   *Cons:*
   * Coarse-grained updates: changes require retagging or creating a new release
   * Updating a single binary necessitates a new full release

While we recognize the strengths of the GitHub Releases model and your valid concerns about reproducibility and version traceability, the team has agreed to proceed with the "Binaries as Regular Git Files" model for now, primarily due to its flexibility and simplicity during the current development phase.

That said, we acknowledge and aim to address the critical issues you raised:

* We introduced explicit versioning for key binary components such as the `xpci` kernel module, to avoid ambiguous or floating references.
* We aligned our repository layout with Marvell’s approach for components like `sai-plugin` to improve clarity and consistency.
* We removed wget calls from .dep files.

We believe these steps will significantly improve reproducibility and maintainability, even within the chosen approach.

mssonicbld commented:
/azp run Azure.sonic-buildimage

aparkhomenko-xsight commented on platform/xsight/kvm-image.mk:
Done in https://github.com/sonic-net/sonic-buildimage/pull/22622/commits/60508346d505f780d173a84bf9c843fb73c73b14 commit.

BorysBarkauskas committed:
[Xsight] fix lspci based Xsight device selection

X2 device name  was changed in lspci output in SONiC 202411

old lspci output:
15:00.0 Non-VGA unclassified device: Device 1e6c:0002 (rev 01)
new lspci output:
15:00.0 Non-VGA unclassified device: Xsight Labs Ltd. XLX2B128A Switch ASIC (rev 01)

using 'lspci -nn' works for both SONiC 202311 and SONiC 202411

Signed-off-by: Borys Barkauskas <bbarkauskas@larch-networks.net>
Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] fix Dockerfile.j2 for syncd docker image

Remove missing unused file from syncd's Dockerfile.j2

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] upgrade XSAI version to 1.16.3-1

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

MichaelZaidman committed:
[ES9618XX] Add build dependency to binary-indep target in debian/rules

The binary-indep target copies compiled kernel modules (.ko files)
into the package but did not depend on the build target that compiles
them. This caused build failures when binary-indep was invoked without
a prior explicit build step.

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>
Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>

MichaelZaidman committed:
[ES9618XX] Port platform kernel modules to Linux 6.12

Adapt ES9618XX platform kernel modules to Linux kernel 6.12 API changes:

- i2c probe callback: use single-argument probe() signature and retrieve
  device identity via i2c_client_get_device_id() to preserve per-device
  driver_data (cpld type, psu index). Without this, CPLD2 sysfs attributes
  are never registered and PSU2 presence/power-good checks use wrong bits.
  This follows the established pattern used by other SONiC platform drivers
  (accton_i2c_cpld, accton_as7712_32x_psu, wistron_sw_to3200k_cpld, etc).
  API changed in kernel 6.3.

- platform_driver remove callback: change return type from int to void.
  API changed in kernel 6.6.

- dps850: remove unused 'chip' field from struct dps850_data.

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>
Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] Use supervisor-proc-exit-listener-rs in syncd container

Alighn Xsight's syncd docker image with latest changes to other docker
images
As done in commit 9b3041fef40a8adcb5f3e57c84ae4e7c067811a1

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

BorysBarkauskas committed:
[Xsight] fix xpci dkms build

XPCI DKMS build started failing beacuase of this warning

Selecting previously unselected package xpci-dkms.
 (Reading database ... 251743 files and directories currently
installed.)
 Preparing to unpack .../xpci-dkms_0.7.0_amd64.deb ...
 Unpacking xpci-dkms (0.7.0) ...
 Setting up xpci-dkms (0.7.0) ...
 Loading new xpci/0.7.0 DKMS files...
 WARNING: No kernel headers were found, skipping module build.
          To get the headers for the running kernel (5.4.0-150-generic)
          please install the linux-headers-5.4.0-150-generic package.

It happens because DKMS takes KVERSION from build environment and
sonic-buildimage build environment has KVERSION 5.4.0-150-generic

So it's needed to run 'dkms install' explicitly during the build

Signed-off-by: Borys Barkauskas <bbarkauskas@larch-networks.net>
Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] fix typo in xpci.dep

xpci.dep had a typo in SPATH what led to a caching error
On older sonic-buildimage versions this error was hidden and
a fresh sonic-builimage revealed this issue

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

aparkhomenko-xsight committed:
[Xsight] do not expose Xsight's debug tcp ports

Xsight exposes some tcp ports from syncd docker container for
debugging
According to the code review it's better to not do it

Signed-off-by: Anton Parkhomenko <aparkhomenko@larch-networks.com>

MichaelZaidman committed:
[ES9618XX] Align es9618xx xplt package with SONiC master Python 3.13

The previous xplt-es9618xx-sonicnos deb was built against Python 3.11.
After SONiC master moved to Python 3.13, that package no longer runs
correctly. Point ACCTON_ES9618XX_XPLT_PLATFORM_MODULE at the py3.13
build (3.4.1).

Signed-off-by: Michael Zaidman <michaelz@xsightlabs.com>
